### DISCUSSION TITLE
Using Platformio and include "AudioLibs/AudioKit.h" in  multiple source files - solved

### DISCUSSION THREAD

**Discussion by ritsch** [pschatzmann/arduino-audiokit#110] [Q&A] [answered]
Thanks for the nice libraries using with Olimex  ESP32-ADF.

In a Platformio project using 'arduino-audiokit-hal' library, when the "AudioLibs/AudioKit.h" header are included in two source files, to access to Audiokit functions there is an Error in linking:

`/home/winfried/.platformio/packages/toolchain-xtensa-esp32/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: .pio/build/serial-upload/src/waa_osc.cpp.o:/home/winfried/repos/AVE/waasaaf/firmware/lib/arduino-audiokit-hal/src/audio_hal/common_defines.h:19: multiple definition of SPI_VSPI; .pio/build/serial-upload/src/waa_main.cpp.o:/home/winfried/repos/AVE/waasaaf/firmware/lib/arduino-audiokit-hal/src/audio_hal/common_defines.h:19: first defined here`

This tells that class SPI_VSPI is implemented twice: at 'arduino-audiokit-hal/src/audio_hal/common_defines.h:19:' in both source file,  so I cannot use the header twice, but I want to access  'kit.volume(vol)' function in both sources.

Trying `#undefine AUDIOKIT_SETUP_SD` coded in `AudioKitSettings.h` did also not work.

What is the best way to include Audiokit in more than one source file ?

mfg winfried

**Comment by pschatzmann** [comment, marked as answer]
I changed the definition to be static. 
This should resolve your issue...

**Reply by ritsch** [reply]
Thanks for fast answer, I did an git pull in the lib.

But now I get:

`
/home/winfried/.platformio/packages/toolchain-xtensa-esp32/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: .pio/build/serial-upload/src/waa_osc.cpp.o:(.bss._ZN11audio_tools17pt_AudioKitStreamE+0x0): multiple definition of  audio_tools::pt_AudioKitStream'; .pio/build/serial-upload/src/waa_main.cpp.o:(.bss._ZN11audio_tools17pt_AudioKitStreamE+0x0): first defined here
/home/winfried/.platformio/packages/toolchain-xtensa-esp32/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: .pio/build/serial-upload/src/waa_osc.cpp.o:(.bss._ZN8audiokit12selfAudioKitE+0x0): multiple definition of  audiokit::selfAudioKit'; .pio/build/serial-upload/src/waa_main.cpp.o:(.bss._ZN8audiokit12selfAudioKitE+0x0): first defined here
`

Maybe I do something wrong, will simplify my project and test again.

**Reply by ritsch** [reply]
Now simpified, but error still exist.

a_main.cpp::

```
#include "Arduino.h"
#include "AudioTools.h"
#include "AudioLibs/AudioKit.h"

AudioInfo info(44100, 2, 16);
AudioKitStream kit;

void set_volume(int vol);

void setup(void) {  
  kit.setVolume(96);
  set_volume(90);
}

// Arduino loop - copy sound to out 
void loop() {
}
```
b.cpp
```
#include "Arduino.h"
#include "AudioTools.h"
#include "AudioLibs/AudioKit.h"
extern AudioKitStream kit;

void set_volume(int vol)
{
    kit.setVolume(96);
}
```
Error like above

```
.. multiple definition of audio_tools::pt_AudioKitStream'; .pio/build/serial-upload/src/a_main.cpp.o:(.bss._ZN11audio_tools17pt_AudioKitStreamE+0x0): first defined here
[...]
 .pio/build/serial-upload/src/b.cpp.o:(.bss._ZN8audiokit12selfAudioKitE+0x0): multiple definition of  audiokit::selfAudioKit'; .pio/build/serial-upload/src/a_main.cpp.o:(.bss._ZN8audiokit12selfAudioKitE+0x0): first defined here
```

PS. Platform.ini:

```
[platformio]
description = "Module: LEMiot waasaaf Module"
src_dir = .

[env]
platform = espressif32
framework = arduino
board = esp32-devkitlipo
build_flags = -I${PROJECT_DIR}/src -DID=0 -DCORE_DEBUG_LEVEL=3
			  -DAUDIOKIT_BOARD=1 -DWORKAROUND_MIC_LINEIN_MIXED=false
lib_deps = 
   https://github.com/pschatzmann/arduino-audio-tools
   https://github.com/pschatzmann/arduino-audiokit-hal

lib_ldf_mode = deep+ ; for audiokit

[env:serial-upload]
monitor_speed = 115200
monitor_filters = esp32_exception_decoder
```

**Comment by pschatzmann** [comment]
I am never using multiple translation units, so I can not guarantee that this is working for all optional components.
I have it corrected now and you need to update both the AudioTools and AudioKit library.

If you find any further issues I suggest that you define the offending variable as static and submit a pull request...

**Reply by ritsch** [reply]
Ok, thank you for the advice, try it.

A solution for me is an interface to audiokit function in one file and use them in others, just some for control over OSC of parameters, so I need only to include audiokit headers once, which is ok for me, but I am not experienced enough for providing a best practice solution.

**Reply by ritsch** [reply]
Ok now it is solved working with the example above, 
( had to update both libs, of course)

Thanks for the help, this let me structure the project in a nicer way.
